### PR TITLE
Wait for k8s registry/crd sync before init

### PR DESCRIPTION
**Pull request by rshriram**
This should hopefully fix https://github.com/istio/istio/issues/12117
Sync the cache before starting the servers for the first time..

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -880,6 +883,8 @@ func (s *Server) createK8sServiceControllers(serviceControllers *aggregate.Contr
 	log.Infof("Primary Cluster name: %s", clusterID)
 	args.Config.ControllerOptions.ClusterID = clusterID
 	kubectl := kube.NewController(s.kubeClient, args.Config.ControllerOptions)
+	dummyStop := make(chan struct{})
+	kubectl.WaitForSync(dummyStop)
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go]: It's not right waiting it here. Because we haven't run the informer.

**Comment by hzxuzhonghu** [comment]
I think we can wait for cache sync
1.  before running http/grpc server.   https://github.com/istio/istio/blob/be16d10ec364ea7c74b0608d7542edc8cd8a2ea6/pilot/pkg/bootstrap/server.go#L1020-L1021
2.  in readiness probe handler
 https://github.com/istio/istio/blob/be16d10ec364ea7c74b0608d7542edc8cd8a2ea6/pilot/pkg/proxy/envoy/v2/debug.go#L506-L513

File: pilot/pkg/config/kube/crd/controller.go [hunk 2]
```diff
@@ -84,11 +84,16 @@ func NewController(client *Client, options kube.ControllerOptions) model.ConfigS
 	for _, schema := range client.ConfigDescriptor() {
 		out.addInformer(schema, options.WatchedNamespace, options.ResyncPeriod)
 	}
+	go out.queue.Run(options.Stop)
```
Comment (costinm) [inline pilot/pkg/config/kube/crd/controller.go]: A small comment will help a lot ( 'start the queue and infromers early' ?)

File: pilot/pkg/config/kube/crd/controller.go [hunk 3]
```diff
@@ -192,32 +197,30 @@ func (c *controller) RegisterEventHandler(typ string, f func(model.Config, model
 	})
 }
 
-func (c *controller) HasSynced() bool {
-	for kind, ctl := range c.kinds {
+func (c *Controller) HasSynced() bool {
+	for _, ctl := range c.kinds {
 		if !ctl.informer.HasSynced() {
-			log.Infof("controller %q is syncing...", kind)
```
Comment (costinm) [inline pilot/pkg/config/kube/crd/controller.go]: Did you add some message when it has synced ?

**Comment by mandarjog** [comment]
I am not convinced that this will make a difference. Readiness check fails when a store is not ready anyway. So regardless of whether the grpc port is open, endpoints should not be added since pilot is not ready.

**Comment by rshriram** [comment]
the CRD is an extra check.. the key thing is the service registry check.. This code starts all the informers, and waits for the caches to sync once before actually starting Pilot. The cache sync shouldn't take much time, as you can see from all the pilot e2e tests that pass (I need to fix the others).

**Comment by duderino** [comment]
The TestTempWorkspace tests are failing in multiple circleci and prow builds.  One example is e2e-pilot-auth-v1alpha3-v2 has this...

https://circleci.com/gh/istio/istio/343802?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

```
TestTempWorkspace - controller
Failed
info	Created namespace istio-test-n5rlx
info	registering for apiVersion networking.istio.io/v1alpha3
info	registering for apiVersion config.istio.io/v1alpha2
info	registering for apiVersion authentication.istio.io/v1alpha1
info	registering for apiVersion rbac.istio.io/v1alpha1
info	registering for apiVersion test.istio.io/v1
info	registering CRD "mockconfigs.test.istio.io"
info	established CRD "mockconfigs.test.istio.io"
```

**Comment by rshriram** [comment]
@morvencao please test with this PR

File: pilot/pkg/config/kube/crd/controller.go [hunk 4]
```diff
@@ -84,11 +84,16 @@ func NewController(client *Client, options kube.ControllerOptions) model.ConfigS
 	for _, schema := range client.ConfigDescriptor() {
 		out.addInformer(schema, options.WatchedNamespace, options.ResyncPeriod)
 	}
+	go out.queue.Run(options.Stop)
+
+	for _, ctl := range out.kinds {
+		go ctl.informer.Run(options.Stop)
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crd/controller.go]: could this cause a race run informer and queue before calling `RegisterEventHandler`

Comment (rshriram) [reply pilot/pkg/config/kube/crd/controller.go]: why is that a race?

**Comment by morvencao** [comment]
@rshriram The good news is that the patch is effective in customer's env. 
Their Istio applications are still working even the pilot is OOM killed.

**Comment by hzxuzhonghu** [comment]
@morvencao  please test #12214, which is a replace

**Pull request closed without merging**